core/src/queue.rs

```diff
@@ -202,20 +239,15 @@ impl Queue {
                 // When transactions are submitted quickly it can be reached.
                 Entry::Vacant(_) => continue,
             };
-            if self.check_tx(entry.get(), wsv).is_err() {
+
+            if !self.is_pending(entry.get(), wsv) {
                 entry.remove_entry();
                 continue;
             }
 
             // Transactions are not removed from the queue until expired or committed
```

Comment:
and why not? Only sumeragi is ever popping transactions out of the queue. It might as well remove them from the queue

Reply:
I think there is 2 reasons: 

1. This way we prevent duplicates from being added into the queue. 
2. Gossiping take transactions from the queue so we want transactions to state here until transactions are in blockchain and we don't want to gossip them anymore.

Reply:
> This way we prevent duplicates from being added into the queue.

how can we have duplicates in the queue when we use Set/Map

> Gossiping take transactions from the queue so we want transactions to state here until transactions are in blockchain and we don't want to gossip them anymore.

gossiping is different I agree, but any time we move transactions into sumeragi cache they should be removed from the queue immediately

Reply:
> how can we have duplicates in the queue when we use Set/Map

If we are talking about removing transactions from the queue then them should be removed from `Map` as well because otherwise we would end up in the situation where `Map` is never cleared. 

> gossiping is different I agree, but any time we move transactions into sumeragi cache they should be removed from the queue immediately

Then we would need to extend gossiping to partially take transactions from sumeragi cache partially from queue which sounds clumsily to me.

Reply:
> Then we would need to extend gossiping to partially take transactions from sumeragi cache partially from queue which sounds clumsily to me.

valid point